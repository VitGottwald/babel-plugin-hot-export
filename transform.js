/** This babel-plugin adds
    `import { hot } from 'react-hot-loader';`

const _Name = hot(module)(Name);
export { _Name as Name };

*/
const p = require("path");

module.exports = function(babel) {
  const { types: t } = babel;

  return {
    name: "Hot Export React Components",
    visitor: {
      Program(path, state) {
        if (
          state.opts.files.every(file => p.resolve(file) !== p.resolve(state.file.opts.filename))
        ) {
          return;
        }
        path.unshiftContainer(
          "body",
          t.ImportDeclaration(
            [t.ImportSpecifier(t.Identifier("hot"), t.Identifier("hot"))],
            t.StringLiteral("react-hot-loader")
          )
        );
      },
      ExportDefaultDeclaration(path, state) {
        if (
          state.opts.files.every(file => p.resolve(file) !== p.resolve(state.file.opts.filename))
        ) {
          return;
        }
        if (t.isClassDeclaration(path.get("declaration"))) {
          const declaration = path.get("declaration").node;
          const identifier = () => declaration.id || t.Identifier("__component");
          declaration.id = identifier();
          path.insertBefore(declaration);
          path.node.declaration = t.CallExpression(
            t.CallExpression(t.Identifier("hot"), [t.Identifier("module")]),
            [identifier()]
          );
        }
      }
    }
  };
};
