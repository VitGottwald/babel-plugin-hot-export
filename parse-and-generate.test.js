import transform from "./parse-and-generate.js";

it("renames varable `n` to `x`", () => {
  const code = `function square(n) {
  return n * n;
}`;
  expect(transform(code)).toMatchSnapshot();
});
