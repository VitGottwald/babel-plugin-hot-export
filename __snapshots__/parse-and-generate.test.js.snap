// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renames varable \`n\` to \`x\` 1`] = `
"function square(x) {
  return x * x;
}"
`;
